proposals/0303-swiftpm-extensible-build-tools.md

```diff
@@ -287,130 +251,224 @@ protocol TargetBuildContext {
 
-- The *commands* that are constructed by an plugin using the `CommandConstructor` API are invoked before, during, or after each build, in accordance with their defined input and output dependencies.
+- The *commands* that are created by an plugin using the `CommandConstructor` API are invoked either before or during the build, in accordance with their defined input and output dependencies.
 
 Binary targets will be improved to allow them to contain command line tools and required support files (such as the system `.proto` files in the case of `protoc`, etc). Binary targets will need to support different executable binaries based on platform and architecture. This is the topic of a separate evolution proposal to extend binary targets to support executables in addition to XCFrameworks.
 
 #### Plugin Capabilities
 
-This proposal defines three different plugin capabilities, which determine how the plugin is used.
+Althought this proposal defines a general approach to plugins, it defines only one initial capability: `buildTool`.  This capability is used for both prebuild commands and build commands, and may in the future be used for additional kinds of build tool invocations (in which varations in build commands would be expressed through new `CommandConstructor` APIs).  Significantly different types of plugin functionality are likely to use different capabilities.
+
+In the package manifest, the capability is expressed a function invocation rather than an enum in order to make it easier to add arguments with default values in the future.
+
+#### Prebuild Command Semantics
+
+Commands created using `addPrebuildCommand()` are run before the start of every build. When creating prebuild commands, the plugin needs to specify a directory into which the outputs will be written.  This is how the prebuild command communicates its outputs to the build system.
+
+Before invoking the prebuild command, the build system will create the output directory if needed, but it will not remove any directory contents that already exist.  After invoking the command, SwiftPM will use the contents of that directory as inputs to the consruction of build commands. The prebuild command should add or remove files so that the directory contents match the source files that should be processed by the build system.
+
+Examples of plugins that need to use prebuild commands include SwiftGen and other tools that need to see all the input files, and whose output files are determined by the contents (not just the paths) of the input files. Such a plugin usually generates just one command and configures it with an output directory into which all generated sources will be written.
+
+Because they run on every build, prebuild commands should use caching to do as little work as possible (ideally none) when there are no changes to their inputs.
```

Comment:
could / should the API for Prebuild Commands include some kind of "Storage" that the plugin can use to save the state to making it easier to compute changes between invocations?

Reply:
Perhaps.  The expectation was that the plugin would use the `outputDirectory` for any cached files between invocations, and that it would pass a command line argument to the prebuild command to a directory inside of its `outputDirectory`.  Perhaps the TargetBuildContext directory needs a better name, since it's easy to confuse with the `outputDirectory` of the prebuild command.

Reply:
I should probably improve one of the examples to illustrate this as well.

Reply:
My understanding was that

```
        /// Optional initial working directory of the command.
         workingDirectory: Path? = nil,
```

can be used for the "skratchpad" for such plugin, say I can write a bunch of timestamps there to know which files I need to scan again etc; the same could be used for some "already generated files" etc right?

Reply:
I added this:
```
    /// A plugin would usually create a separate subdirectory of this directory
    /// for each command it creates, and the command would be configured to
    /// write its outputs to that directory. The plugin may also create other
    /// directories for cache files and other file system content that either
    /// it or the command will need.
```
but will also see if I can extend one of the examples.

Reply:
I changed the name to `pluginWorkDirectory` to differentiate it from the command output directories, but am not overjoyed with that name.  Open to suggestions.  I'm sure there will be plenty of naming opinions during review and this can easily be adjusted if the semantics stay the same.